### PR TITLE
Sort by

### LINKED ISSUE #41: Suggestion: sorting of displayed cards
It would be useful to be able to sort the displayed cards by:

- name
- reference (JU001, OP034, etc.)
- last modified

### PR DESCRIPTION
**Pull request by goulinkh**
## Done

- Add sortable cards, sort by: "name", "reference' and "last updated"

## QA

- Go to: https://specs-canonical-com-50.demos.haus/
- Select on of the 3 sort options
- Make sure that the cards are sorted by the selected option

## Issues

Fixes #41

### CODE REVIEW

**Comment by tbille** [comment]
As discussed in standup, order should be: `"last updated", "name", "reference"`

**Comment by tbille** [comment]
@goulinkh sorry I missed the sort dropdown. By default last modified should be selected. There should be only 3 options: `"last updated", "name", "reference"`

What do you mean by reference? Is it the spec id?

And finally can the sort be at the bottom, by default we want users to look at the latest modified specs, the chances of changing the sort over searching for a team are lower.

**Comment by goulinkh** [comment]
> What do you mean by reference? Is it the spec id?

@tbille Yes! reference is index-id, this maybe should be named `reference (index)`?

**Comment by tbille** [comment]
@goulinkh can you rename `reference` to `Spec index` instead?

File: templates/_base.html [hunk 1]
```diff
@@ -88,6 +83,26 @@
         "author"
       );
 
+      function sortCards(by = "name") {
+        let key = () => "";
+        if (by === "name") {
+          key = (e) => e.dataset.title;
```
Comment (andesol) [inline templates/_base.html:90]: Maybe this could be "card" instead of "e"? (since e sounds like "event")

File: templates/_base.html [hunk 2]
```diff
@@ -88,6 +83,26 @@
...
+        } else if (by === "index") {
+          key = (e) => e.dataset.index;
+        } else if (by === "last-updated") {
+          key = (e) => new Date(e.dataset.lastupdated).getTime();
+        }
+        const cards = document.getElementById("cards");
+        Array.from(cards.children)
+          .map((card) => cards.removeChild(card))
```
Comment (andesol) [inline templates/_base.html:99]: This could be a `forEach`, since we are mutating something, rather than generating a new array

Comment (goulinkh) [reply templates/_base.html:99]: `removeChild` returns an element that we need later in the next steps `sort` and `forEach`

**Comment by andesol** [comment]
@goulinkh When I sort by name, the cards don't look like alphabetically sorted. Or maybe I'm missing something.

File: templates/_base.html (hunk 2 above)
Comment (goulinkh) [reply templates/_base.html:99]: > @goulinkh When I sort by name, the cards don't look like alphabetically sorted. Or maybe I'm missing something.

thanks @andesol for the QA, fixed that issue

**Comment by goulinkh** [comment]
> @goulinkh When I sort by name, the cards don't look like alphabetically sorted. Or maybe I'm missing something.

Thanks @andesol for the QA, fixed the mentioned issue

**Comment by andesol** [comment]
@goulinkh I just noticed sorting by spec index that some cards have "unknown" in their heading but they actually have an index set in `data-index` (and thus the sorting works fine). This is due to a seemingly "random" condition that limits the index to five characters: https://github.com/goulinkh/specs.canonical.com/blob/sort-by/templates/_card.html#L17

Do you know what's the reasoning behind this?

File: templates/_base.html (hunk 2 above)
Comment (andesol) [reply templates/_base.html:99]: Oh, now I get it

**Comment by goulinkh** [comment]
@andesol I think it's part of the spec template definition https://docs.google.com/document/d/160ACH_cCB95oPtU0jN6dE797GdzbEtvy1N50XhPRtCY/edit

**Comment by andesol** [comment]
> @andesol I think it's part of the spec template definition https://docs.google.com/document/d/160ACH_cCB95oPtU0jN6dE797GdzbEtvy1N50XhPRtCY/edit

But then the index is not "unknown", it just doesn't conform to the specified format. It seems inconsistent to not display it but use it for sorting. 

But I guess we can open a new issue for this, since the sorting works fine.

**Review by andesol — APPROVED** [review]

**Pull request merged**